doc for deploying compound solidity and uniswap vyper (#96)
oasislabs/docs: src/tutorials/deploy-vyper.md

# Deploying Vyper Contracts on Oasis (Uniswap)

> [Uniswap](https://uniswap.io/) is a decentralized exchange (DEX) for ERC20 tokens built on Ethereum. Here, it is configured for deployment on Oasis. This tutorial's methods for deploying Solidity contracts using Truffle and Web3.js are applicable to any Ethereum contract written in Solidity. For contracts written in Vyper, see the Uniswap tutorial.

Reviewer: I think the last 1-2 sentences of this still refer to Compound not Uniswap

## Quickstart

> If you have the Oasis Toolchain installed and just want to deploy right away:

Reviewer: ```suggestion
If you have the Oasis toolchain installed and just want to deploy right away:
```

1. Clone this repo.
2. Install dependencies with `npm install`.
3. Run a local oasis network with `oasis chain` in a separate terminal window.
4. Run `node deploy-uniswap` to deploy the contracts.

## Prerequisites

> 

Reviewer: Remember to make same changes to prerequisites here that you make above :)

2019-09-27 10:54:21,492 INFO  [oasis_chain] Oasis local chain is running
```

Save the Mnemonic; you will need it later. 

Note that the port :8546 corresponds to ws. You will need to use the port :8545 for http connections. 
You might also want to save one of the public addresses to serve as your account to deploy contracts with. For example you can copy-paste an account like so:

```js
my_address = '0xb8b3666d8fea887d97ab54f571b8e5020c5c8b58'
```

## Compiling the Contracts

### Install Vyper

Clone the [Vyper repository](https://github.com/ethereum/vyper). 
```bash 
git clone https://github.com/ethereum/vyper
cd vyper
```
To retrieve the version that the Uniswap contracts are compatible with, you should reset to a previous version. If you don't have the correct version, you may get compiler errors. 
```bash
git reset --hard 35038d20bd9946a35261c4c4fbcb27fe61e65f78
```

Vyper requires Python version 3.6 or higher. The best way to install vyper is to use `virtualenv` to create a python virtual environment:

First make sure you have Python 3.6 installed, then get your path to python 3.
```bash
which python3
```
Start a virtualenv.
```

> virtualenv -p <PATH/TO/PYTHON3.6> my_project

Reviewer: ```suggestion
virtualenv -p /path/to/python3.6 my_project
```

source my_project/bin/activate
```
Then, run `make` in your `vyper` directory. You may need to use `sudo` if on mac:
```bash
sudo make
```
Vyper should be installed now and you should be able to compile contracts. The [Vyper Documentation](https://vyper.readthedocs.io/en/latest/installing-vyper.html) has more installation information.

> Uniswap Vyper contracts can be found at [this repository](https://github.com/Uniswap/contracts-vyper). 

Reviewer: ```suggestion
Uniswap Vyper contracts can be found [here](https://github.com/Uniswap/contracts-vyper). 
```

> You can compile using the `vyper` command or using truffle. 

Reviewer: ```suggestion
You can compile using the `vyper` command or using Truffle. 
```

Reviewer: Or `truffle`

## Deploying the Contracts

### Deploy Using Truffle

> If you choose to deploy using truffle, you'll need to create migrations contracts. 

Reviewer: ```suggestion
If you choose to deploy using Truffle, you'll need to create migrations contracts. 
```

The first are a `Migrations.sol` in your project's `./contracts` folder and `1_initial_migrations` in your project's `./migrations` folder, both of which can be automatically generated using `truffle init`. 

You also need a `2_deploy_contracts` which should look like this:
```js
var uniswap_exchange = artifacts.require("./uniswap_exchange"); //note there is no ".vy"
var uniswap_factory = artifacts.require("./uniswap_factory"); //note there is no ".vy"

module.exports = function(deployer, network, accounts) {
  deployer.deploy(uniswap_exchange, accounts[0]);
  deployer.deploy(uniswap_factory);
}
```

> Use `truffle migrate` to deploy.

Reviewer: ```suggestion
Run `truffle migrate` to deploy.
```

### Deploy Using Web3

> First, connect to Web3 using a `truffle-hdwallet-provider`. You can run `oasis-chain` to start up a local network and retrieve your mnemonic and url. 

Reviewer: ```suggestion
First, connect to Web3 using a `truffle-hdwallet-provider`. You can run `oasis chain` to start up a local blockchain server and retrieve your mnemonic and URL. 
```

```js
const Web3 = require('web3');
const HDWalletProvider = require("truffle-hdwallet-provider");
const MNEMONIC = 'range drive remove bleak mule satisfy mandate east lion minimum unfold ready';
const URL = 'http://localhost:8545';
const provider = new HDWalletProvider(MNEMONIC, URL);
const web3 = new Web3(provider);
```

Using the contract ABI and bytecode from the compilation of your vyper contracts, instantiate new contract objects. Uniswap requires a **factory** contract to create new exchanges, and an **exchange** contract that serves as a template for each exchange.

```js
factory_json = fs.readFileSync('./path/to/file.json', 'utf8');
factory_abi = JSON.parse(factory_json)["abi"];
factory_bytecode = JSON.parse(factory_json)["bytecode"];
exchange_json = fs.readFileSync('./path/to/file.json', 'utf8');
exchange_abi = JSON.parse(exchange_json)["abi"];

> exchange_bytecode = JSON.parse(exchange_json)["bytecode"];

Reviewer: ```suggestion
const factory_json = fs.readFileSync('./path/to/file.json', 'utf8');
const factory_abi = JSON.parse(factory_json)["abi"];
const factory_bytecode = JSON.parse(factory_json)["bytecode"];
const exchange_json = fs.readFileSync('./path/to/file.json', 'utf8');
const exchange_abi = JSON.parse(exchange_json)["abi"];
const exchange_bytecode = JSON.parse(exchange_json)["bytecode"];
```

```

Initialize your contract objects.

```js
factory_contract = new web3.eth.Contract(factory_abi);

> exchange_contract = new web3.eth.Contract(exchange_abi);

Reviewer: const factory_contract = new web3.eth.Contract(factory_abi);
const exchange_contract = new web3.eth.Contract(exchange_abi);

```

Finally, deploy your contracts and save their addresses for future use.

```js
factory_contract.deploy({data:factory_bytecode})
.send( { from: my_address }) 
.on('error', function(err){ console.log("error: " + err)})
.then(function(newContractInstance){
    console.log("Factory Deployed successfully:" + newContractInstance.options.address);
    factory_contract.options.address = newContractInstance.options.address;
});
exchange_contract.deploy({data:exchange_bytecode})
.send( { from: my_address }) 
.on('error', function(err){ console.log("error: " + err)})
.then(function(newContractInstance){
    console.log("Exchange Deployed successfully:" + newContractInstance.options.address);
    exchange_contract.options.address = newContractInstance.options.address;

> });

Reviewer: factory_contract.deploy({
    data: factory_bytecode,
}).send({
    from: my_address,
}).on('error', (err) => {
    console.log(`Deploy failed with error: ${err}`)
}).then((deployment) => {
    console.log(`Factory deployed successfully, at address: ${deployment.options.address}`);
    factory_contract.options.address = deployment.options.address;
});
exchange_contract.deploy({
    data: exchange_bytecode,
}).send({
    from: my_address,
}).on('error', (err) => {
    console.log(`Deploy failed with error: ${err}`)
}).then((deployment) => {
    console.log(`Exchange Deployed successfully: ${deployment.options.address}`);
    exchange_contract.options.address = deployment.options.address;
});

```

> The HD Wallet Provider may continue running and cause execution to "hang" at the end. To gracefully exit, use the stop function:

Reviewer: ```suggestion
The HD Wallet Provider may continue running and cause execution to hang at the end. To gracefully exit, use the stop function:
```

```js
provider.engine.stop();
```

## Testing the Deployment

Once you've deployed your contracts, you can interact with them! Try initializing the factory by giving it a template (the exchange contract address). Make sure to do all of this before you call the stop function.

Assuming you have your factory and exchange contracts already deployed, fill these in with their corresponding addresses:

```js

> factory_address = '0x...'; 

Reviewer: `const` before variable names

exchange_address = '0x...';
factory_contract = new web3.eth.Contract(factory_abi, factory_address);
exchange_contract = new web3.eth.Contract(exchange_abi, exchange_address);
```

To initialize the factory:

```js

> factory_contract.methods.initializeFactory(exchange_address).send( {from: my_address}).then( () => {

Reviewer: ```suggestion
factory_contract.methods.initializeFactory(exchange_address).send({
    from: my_address,
}).then(() => {
    factory_contract.methods.exchangeTemplate().call((err, result) => {
        console.log(`\n\nSuccessfully initialized Uniswap factory with template: ${result}`);
    });
}).catch(console.log);
```

    factory_contract.methods.exchangeTemplate().call((err, result) => {
      console.log("\n\nSuccessfully initialized Uniswap factory with template: " + result)
    });
  }).catch(console.log);
```

You can also create a new exchange for the ERC20 token of your choice. All you need is the address to the token contract.

```js
token_address = '0x...';
factory_contract.methods.createExchange(token_address).send( {from: my_address}).then(() => {
    factory_contract.methods.getExchange( token_contract.options.address).call((err, result) => {
      console.log("Successfully created a new exchange for token at " + result);
    });

>   })

Reviewer: ```suggestion
const token_address = '0x...';
factory_contract.methods.createExchange(token_address).send({
    from: my_address
}).then(() => {
  factory_contract.methods.getExchange( token_contract.options.address).call((err, result) => {
      console.log(`Successfully created a new exchange for token at ${result}`);
  });
})
```

```

Now you can deploy and interact with any Vyper contract on the Oasis platform!

> 

Reviewer: Remove extra newline
